docs/design-documents/tools/cmake_components.md

```diff
@@ -45,9 +45,32 @@ targets:
 
 Breaking the dependencies would be a huge effort and it possibly would result in rewriting these components as they were for years considered as monolitic - having everything available.
 
-## Components as object libraries
+## Components libraries
 
-CMake provides OBJECT libraries but it does not support circular dependencies that we have in our tree. Therefore we build Mbed OS as whole (all object files combined).
+Mbed OS uses various type of CMake libraries:
+
+`STATIC` libraries should be the default library type used to build Mbed OS CMake
+libraries. However, due to its inability to resolve weakly linked
+symbols with strong ones, we have to use other types when a library
+contains weak symbols.
+
+`OBJECT` libraries are used in Mbed OS to generates objects file but not archiving
+them. It has the advantage of not always rebuilding files if there aren't any
+changes and can resolve weakly linked symbols with strong ones. Its main
+disadvantage is that it does not allow circular dependency between libraries.
+
+`INTERFACE` libraries are used to model usage requirements for a target that is
+outside our project. It is possible to use it with libraries that have circular
+depencies, however, interface libraries produce no output in in the build system.
+
+Note:
+Include directories of `INTERFACE` libraries privately linked with `OBJECT` libraries
```

Comment:
This is the use case we had but it should be true to any library type  - how private linking works for any library. However, this is changes when you use OBJECT libraries. As if they are linked publicly, the objects are shared. See https://cmake.org/cmake/help/latest/command/target_link_libraries.html#id7 for more details. We do not want to share objects between libraries as this leads to duplicate definitions found error.

I would word this something like:

If a library contains public headers (for instance a driver configuration that extends stack configuration, this configuration should be exposed to an application), they should be shared via PUBLIC attribute in target_include_directories. Due to OBJECT libraries sharing objects when linked publicly, we link privately in some cases to not share object files but we need to expose their headers. This can be achieved by using the `PUBLIC` attribute and the generator expression `$<TARGET_PROPERTY:mbed-interface-library>,INTERFACE_INCLUDE_DIRECTORIES>`